x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticMatchQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,106 @@
+        if (resolvedIndices != null) {
+            Collection<IndexMetadata> indexMetadataCollection = resolvedIndices.getConcreteLocalIndicesMetadata().values();
+            List<String> inferenceIndices = new ArrayList<>();
+            List<String> nonInferenceIndices = new ArrayList<>();
+            for (IndexMetadata indexMetadata : indexMetadataCollection) {
+                String indexName = indexMetadata.getIndex().getName();
+                InferenceFieldMetadata inferenceFieldMetadata = indexMetadata.getInferenceFields().get(matchQueryBuilder.fieldName());
+                if (inferenceFieldMetadata != null) {
+                    inferenceIndices.add(indexName);
+                } else {
+                    nonInferenceIndices.add(indexName);
+                }
+            }
+
+            if (inferenceIndices.isEmpty()) {
+                return rewritten;
+            } else if (nonInferenceIndices.isEmpty() == false) {
+                BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+                for (String inferenceIndexName : inferenceIndices) {
+                    // Add a separate clause for each semantic query, because they may be using different inference endpoints
+                    // TODO - consolidate this to a single clause once the semantic query supports multiple inference endpoints
+                    boolQueryBuilder.should(
+                        createSemanticSubQuery(inferenceIndexName, matchQueryBuilder.fieldName(), (String) matchQueryBuilder.value())
+                    );
+                }
+                boolQueryBuilder.should(createMatchSubQuery(nonInferenceIndices, matchQueryBuilder.fieldName(), matchQueryBuilder.value()));
+                rewritten = boolQueryBuilder;
+            } else {
+                rewritten = new SemanticQueryBuilder(matchQueryBuilder.fieldName(), (String) matchQueryBuilder.value(), false);
+            }
+        }
+
+        return rewritten;
+
+    }
+
+    @Override
+    public String getName() {
+        return MatchQueryBuilder.NAME;
+    }
+
+    private QueryBuilder createSemanticSubQuery(String indexName, String fieldName, String value) {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must(new SemanticQueryBuilder(fieldName, value, true));
+        boolQueryBuilder.filter(new TermQueryBuilder(IndexFieldMapper.NAME, indexName));
+        return boolQueryBuilder;
+    }
+
+    private QueryBuilder createMatchSubQuery(List<String> indices, String fieldName, Object value) {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must(new AbstractSemanticMatchQueryWrapper(fieldName, value));
+        boolQueryBuilder.filter(new TermsQueryBuilder(IndexFieldMapper.NAME, indices));
+        return boolQueryBuilder;
+    }
+
+    static class AbstractSemanticMatchQueryWrapper extends AbstractQueryBuilderWrapper<MatchQueryBuilder> {
```

Comment:
This is not needed, this wrapping should be handled by the `AbstractQueryBuilder` directly. We don't want to wrap the match query, we want to wrap the query that the interceptor rewrites into. 
So `AbstractQueryBuilder#rewrite` could do something like:
```
if (interceptor != null) {
   var rewritten = interceptor.rewrite(this);
   if (rewritten != this) {
      return new InterceptQueryWrapper(rewritten);
   }
```
and then the interceptor can rewrite into any query without bothering about this wrapping.

Reply:
The issue with your suggested approach is the fact that `rewrite` returns a `QueryBuilder` and not an `AbstractQueryBuilder`. We need to intercept at the `AbstractQueryBuilder` level so we have access to `doRewrite` and it feels like this optimization may not be worth it, when we can just wrap it during interception.

Reply:
> The issue with your suggested approach is the fact that rewrite returns a QueryBuilder and not an AbstractQueryBuilder. 

We're calling rewrite and not `doRewrite` so not sure what's preventing this change. Also I think you missed the fact that we don't want to wrap the match query builder, we want to wrap any query builder that is returned by the rewrite call.